core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -306,14 +297,10 @@ private[spark] class ExternalSorter[K, V, C](
 
     var success = false
     try {
-      val it = collection.destructiveSortedIterator(partitionKeyComparator)
+      val it = collection.destructiveSortedWritablePartitionedIterator(comparator)
       while (it.hasNext) {
-        val elem = it.next()
-        val partitionId = elem._1._1
-        val key = elem._1._2
-        val value = elem._2
-        writer.write(key)
-        writer.write(value)
+        val partitionId = it.nextPartition()
+        it.writeNext(writer)
         elementsPerPartition(partitionId) += 1
         objectsWritten += 1
```

Comment:
I suspect that this patch has rendered `spark.shuffle.spill.batchSize` useless.  It looks like we don't have any tests that are capable of detecting whether this setting actually takes effect or not.